BUG: Fix to_json lines with escaped characters

rouzazari opened the pull request:
Updates existing to_json methodology by adding is_escaping variable,
which ensures escaped chars are handled correctly.

Bug description: A simple check of whether the prior char is a backslash
is insufficient because the backslash may itself be escaped.

A test is also included (previously included in #14693).

xref #14693
xref #15096

 - [x] closes #15096 (#14693 was previously closed by author)
 - [x] tests added / passed
 - [x] passes ``git diff upstream/master | flake8 --diff`` (used ``pep8radius master --diff`` instead)
 - [x] whatsnew entry (does not seem necessary, but let me know if I should add)

rouzazari committed:
BUG: Fix to_json lines with escaped characters

Updates existing to_json methodology by adding is_escaping variable, which ensures escaped chars are handled correctly.

- Includes test for escaped characters in keys and values (i.e. columns and data).
- Includes bug fix in whatsnew
- Revised type of in_quotes and is_escaping to bint

xref #14693
xref #15096

jreback commented on pandas/lib.pyx:
```diff
@@ -1099,6 +1099,7 @@ def convert_json_to_lines(object arr):
     """
     cdef:
         Py_ssize_t i = 0, num_open_brackets_seen = 0, in_quotes = 0, length
+        Py_ssize_t is_escaping = 0
```
make in_quotes and is_escaping of type ``bint``

jreback commented:
pls add a whatsnew note (0.20.0 bug fix section)

rouzazari commented:
Thanks, @jreback. I made the changes you requested, rebased to master, and squashed the commits.

Please note: I revised the test to include a test for escaped characters in keys and values (i.e. columns and data of dataframe). Previously I was only testing escaped characters in values/data.

jreback commented on doc/source/whatsnew/v0.20.0.txt:
```diff
@@ -351,6 +351,7 @@ Bug Fixes
 - Require at least 0.23 version of cython to avoid problems with character encodings (:issue:`14699`)
 - Bug in converting object elements of array-like objects to unsigned 64-bit integers (:issue:`4471`, :issue:`14982`)
 - Bug in ``pd.pivot_table()`` where no error was raised when values argument was not in the columns (:issue:`14938`)
+- Bug in ``.to_json()`` where lines=True and contents (keys or values) contain escaped characters (:issue:`15096`)
```
use double-backticks around ``lines=True``

jreback commented on pandas/io/tests/json/test_pandas.py:
```diff
@@ -972,6 +972,13 @@ def test_to_jsonl(self):
         self.assertEqual(result, expected)
         assert_frame_equal(pd.read_json(result, lines=True), df)
 
+        # GH15096
```
add a 1-line comment on what you are testing here

jreback commented:
lgtm, minor comments and some linting issues to fix. ping on green.

jreback commented:
thanks! nice concise PR! keep em coming

rouzazari commented:
Thanks, @jreback. I appreciate the reviews and feedback.

FYI: There is no change in runtime performance from the bug fix. This is somewhat obvious, but it was a good exercise for me, so I thought I'd share. I reran the following performance tests from #14408:

Master:
```
In [5]: df = DataFrame(dict([('float{0}'.format(i), np.random.randn(N)) for i in range(C)]))
In [6]: %timeit df.to_json('foo.json',orient='records',lines=True)
10 loops, best of 3: 135 ms per loop
```

This branch:
```
In [5]: df = DataFrame(dict([('float{0}'.format(i), np.random.randn(N)) for i in range(C)]))
In [6]: %timeit df.to_json('foo.json',orient='records',lines=True)
10 loops, best of 3: 135 ms per loop
```

jreback commented:
@rouzazari great!

note that we already have some asv's for perf testing, I am not sure how much they cover for json though, you can have a look in pandas/asv/benchmarks/packers.py

if you think some cases need more, then please PR!

rouzazari commented:
Got it. I'm not too familiar with this area, but I'll take a look.
